crates/ruff_server/src/server/api/requests/code_action.rs

```diff
@@ -1,81 +1,222 @@
+        // fast path - return early if no actions are available
+        if available_actions.is_empty() {
+            return Ok(None);
+        }
         let document = snapshot.document();
-        let url = snapshot.url();
-        let encoding = snapshot.encoding();
-        let version = document.version();
-        let actions: Result<Vec<_>> = params
-            .context
-            .diagnostics
-            .into_iter()
-            .map(|diagnostic| {
-                let Some(data) = diagnostic.data else {
-                    return Ok(None);
-                };
-                let diagnostic_fix: crate::lint::DiagnosticFix = serde_json::from_value(data)
-                    .map_err(|err| anyhow::anyhow!("failed to deserialize diagnostic data: {err}"))
-                    .with_failure_code(lsp_server::ErrorCode::ParseError)?;
-                let edits = diagnostic_fix
-                    .fix
-                    .edits()
-                    .iter()
-                    .map(|edit| types::TextEdit {
-                        range: edit.range().to_range(
-                            document.contents(),
-                            document.index(),
-                            encoding,
-                        ),
-                        new_text: edit.content().unwrap_or_default().to_string(),
-                    });
-
-                let changes = vec![types::TextDocumentEdit {
-                    text_document: types::OptionalVersionedTextDocumentIdentifier::new(
-                        url.clone(),
-                        version,
-                    ),
-                    edits: edits.map(types::OneOf::Left).collect(),
-                }];
-
-                let title = diagnostic_fix
-                    .kind
-                    .suggestion
-                    .unwrap_or(diagnostic_fix.kind.name);
-                Ok(Some(types::CodeAction {
-                    title,
-                    kind: Some(types::CodeActionKind::QUICKFIX),
+        // compute the associated document edits for each diagnostic
+        // these will get re-used when building the actual code actions afterwards
+        let edits = diagnostic_edits(
+            document,
+            snapshot.url(),
+            snapshot.encoding(),
+            document.version(),
+            params.context.diagnostics.into_iter(),
+        )
+        .collect::<Result<Vec<_>>>()?;
+
+        let mut response: types::CodeActionResponse = types::CodeActionResponse::default();
+
+        for action in available_actions {
+            match action {
+                SupportedCodeAction::QuickFix => response.extend(quick_fix(edits.iter())),
+                SupportedCodeAction::FixAll => response.extend(fix_all(edits.iter())),
+                SupportedCodeAction::OrganizeImports => {
+                    todo!("Implement the `source.organizeImports` code action")
+                }
+            }
+        }
```

Comment:
Nit: Considering that `available_actions` is a set (or bitset, see below), I would probably rewrite it as

```
			  // Add a quick fix for the diagnostic
        if available_actions.contains(&SupportedCodeAction::QuickFix) {
            response.extend(quick_fix(edits.iter()));
        }

				// Add a single fix all action for all diagnostics
        if available_actions.contains(&SupportedCodeAction::FixAll) {
            response.extend(fix_all(edits.iter()));
        }

        if available_actions.contains(&SupportedCodeAction::OrganizeImports) {
            todo!("Implement the `source.organizeImports` code action")
        }
```

I find this more readable because it makes the ordering explicit in code (no need to implement `Ord`) and allows for inline comments.